Subject:       Changes to boot from plain dm-crypt device (v2)
Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch Status:  Complete on 3.5.x.x
Date:          2019-01-28 05:01:05.478991171 -0800

v2 - simplify conditional
v1 - initial release

diff -urp genkernel-3.5.3.3.orig/defaults/initrd.scripts genkernel-3.5.3.3/defaults/initrd.scripts
--- genkernel-3.5.3.3.orig/defaults/initrd.scripts	2018-05-07 17:22:32.000000000 -0700
+++ genkernel-3.5.3.3/defaults/initrd.scripts	2021-07-19 13:10:34.189571893 -0700
@@ -1195,7 +1195,7 @@ startVolumes() {
 			then
 				good_msg "ZFS pool ${ZFS_POOL} already imported."
 
-				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" ]
+				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 				then
 					good_msg "LUKS detected. Reimporting ${ZFS_POOL}"
 					/sbin/zpool export -f "${ZFS_POOL}"
@@ -1319,8 +1319,7 @@ openLUKS() {
 			LUKS_DEVICE=$(find_real_device "${LUKS_DEVICE}")
 
 			setup_md_device ${LUKS_DEVICE}
-			cryptsetup isLuks ${LUKS_DEVICE}
-			if [ $? -ne 0 ]
+			if [ ! cryptsetup isLuks ${LUKS_DEVICE} -a -z "${CRYPT_ROOT_PLAIN}" ]
 			then
 				bad_msg "The LUKS device ${LUKS_DEVICE} does not contain a LUKS header" ${CRYPT_SILENT}
 				DEV_ERROR=1
@@ -1405,7 +1404,12 @@ openLUKS() {
 					fi
 				fi
 				# At this point, keyfile or not, we're ready!
-				crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				if [ -n "${CRYPT_ROOT_PLAIN}" ]
+				then
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} plainOpen ${CRYPT_ROOT_PLAIN} ${LUKS_NAME}"
+				else
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				fi
 				crypt_filter_ret=$?
 
 				touch ${flag_opened}
@@ -1443,7 +1447,7 @@ startLUKS() {
 	[ -n "${CRYPT_ROOT_KEY}" ] && [ -z "${CRYPT_ROOT_KEYDEV}" ] \
 		&& sleep 6 && bootstrapKey "ROOT"
 
-	if [ -n "${CRYPT_ROOT}" ]; then
+	if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_ROOT_PLAIN}" ]; then
 		openLUKS "root"
 		if [ -n "${REAL_ROOT}" ]
 		then
diff -urp genkernel-3.5.3.3.orig/defaults/linuxrc genkernel-3.5.3.3/defaults/linuxrc
--- genkernel-3.5.3.3.orig/defaults/linuxrc	2018-05-07 17:22:32.000000000 -0700
+++ genkernel-3.5.3.3/defaults/linuxrc	2021-07-19 13:07:40.668491886 -0700
@@ -216,6 +216,9 @@ do
 		crypt_root=*)
 			CRYPT_ROOT=${x#*=}
 		;;
+		crypt_root_plain=*)
+			CRYPT_ROOT_PLAIN=${x#*=}
+		;;
 		crypt_swap=*)
 			CRYPT_SWAP=${x#*=}
 		;;
